facebook/OpenBIC
# fby3.5: cl:Support IPMI command - Get card type

Summary:
- Support 1OU card type detection (by SB BIC firmware)
  Follow hardware design, BIC can judge 1OU card type according the ADC-6(0-based) voltage if 1OU is present.
  BIC should set the type to SB CPLD register "1OU Card Detection" to let SB CPLD powers on the SI test card.
- Support IPMI command - "Get card type"
  NetFn: 38h / Cmd: A1h
  Request Data:
    Byte[3:1] - Manufacturer ID
    Byte4 - Get 1OU/2OU card type(00h for 1OU and 01h for 2OU)
  Response Data:
    Byte1 - Completion code
            00h - Successfully
            C7h - Request data length invalid
            CCh - Invalid data field in Request
    Byte2 - 1OU/2OU(00h for 1OU and 01h for 2OU)
    Byte3 - card type
            1OU -
                * 00h: 1OU SI test card
	        * 01h: Expansion with 6 M.2
	        * 02h: Rainbow falls ( CXL with 4 DDR4 DIMMs)
	        * 03h: Vernal falls (with TI chip)
	        * 04h: Vernal falls (with AST1030 chip)
	        * 05h: Kahuna Falls
	        * 06h: Waimano falls (CXL with 2 DDR4 DIMMs+2 E1S)
	        * 07h: Expansion with NIC
                * FEh: 1OU card is absent
                * FFh: Unknown 1OU card type
            2OU -
                * 07h: DPV2 x8
                * 70h: DPV2 x16
                * FEh: 2OU card is absent
                * FFh: Unknown 2OU card type

Dependency: #303 

TODO:
The manufacture ID should be changed to Meta's IANA.

Test Plan:
1. Check the IPMI command "Get card type" - pass
   ==> Get 1OU card type: bic-util ${slot} 0xe0 0xa1 0x9c 0x9c 0x0 0x0
   ==> Get 2OU card type: bic-util ${slot} 0xe0 0xa1 0x9c 0x9c 0x0 0x1
2. Check SB CPLD register 09h - pass
   ==> bic-util ${slot} 0x18 0x52 0x1 0x42 0x1 0x9
Log:
[1OU - SI test card]
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x0
9C 9C 00 00 00
[1OU - Rainbow falls]
root@bmc-oob:~# bic-util slot2 0xe0 0xa1 0x9c 0x9c 0x0 0x0
9C 9C 00 00 02
root@bmc-oob:~# bic-util slot2 0x18 0x52 0x1 0x42 0x1 0x9
02
[2OU - DPV2 x8]
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x1
9C 9C 00 01 07
[2OU - DPV2 x16]
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x1
9C 9C 00 01 70
[2OU - DPV2 x8 and DPV2 x16]
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x1
9C 9C 00 01 77
[No 1OU/2OU card]
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x0
9C 9C 00 00 FE
root@bmc-oob:~# bic-util slot1 0xe0 0xa1 0x9c 0x9c 0x0 0x1
9C 9C 00 01 FE
